### PR TITLE
Android: Pass GAPID argument --gapir-args from gapis to gapir

### LINKED ISSUE #2839: GAPIR on Android cannot receive arguments
GAPIR can take some arguments on desktop, but there is currently no way to pass arguments on Android.

The solution needs to use JNI to retrieve arguments from the Android intent "extra" field.

Assigned to @purvisa-at-google-com as a first task.

### PR DESCRIPTION
**Pull request by purvisa-at-google-com**
Fix #2839 
This pull request also unifies the gapir logging for PC and Android and adds support for some arguments that were previously not implemented by gapir on Android.

### CODE REVIEW

File: cmd/gapir/cc/main.cpp [hunk 1]
```diff
@@ -76,6 +77,235 @@ std::vector<uint32_t> memorySizes {
...
+      } else if (strcmp(argv[i], "-h") == 0 || strcmp(argv[i], "-help") == 0 ||
+                 strcmp(argv[i], "--help") == 0) {
+        opts.help = true;
+      } else {
+        GAPID_FATAL("Unknown argument: %s", argv[i]);
+      }
+    }
+
+#if TARGET_OS == GAPID_OS_ANDROID
+    if (opts.android) {
```
Comment (AWoloszyn) [inline cmd/gapir/cc/main.cpp]: I am not sure what `--android` is supposed to represent.
It looks like it is supposed to be the equivalent to "--replay-archive" on desktop based on the next little bit of code.

File: core/cc/log.h [hunk 2]
```diff
@@ -132,6 +98,14 @@ class Logger {
 
 }  // namespace core
 
+#if TARGET_OS == GAPID_OS_ANDROID
+#include <android/log.h>
+#else  // TARGET_OS == GAPID_OS_ANDROID
+#include <stdio.h>
+#endif  // TARGET_OS == GAPID_OS_ANDROID
```
Comment (ben-clayton) [inline core/cc/log.h]: Can these includes now be moved to the .cpp?

File: gapir/client/client.go [hunk 3]
```diff
@@ -30,7 +30,7 @@ type tyLaunchArgsKey string
 
 // LaunchArgsKey is the bind device property key used to control the command
 // line arguments when launching GAPIR. The property must be of type []string.
-const LaunchArgsKey tyLaunchArgsKey = "<gapir-launch-args>"
+const LaunchArgsKey tyLaunchArgsKey = "gapir-launch-args"
```
Comment (ben-clayton) [inline gapir/client/client.go]: What's the purpose of this change? This constant is used as a property key, and there's no documentation / any obvious relation to the actual command line parameter name.

File: gapir/client/session.go [hunk 4]
```diff
@@ -295,10 +296,19 @@ func (s *session) newADB(ctx context.Context, d adb.Device, abi *device.ABI) err
 		return err
 	}
 
+	var completeLaunchArgs []string;
```
Comment (ben-clayton) [inline gapir/client/session.go]: no semicolons. Please use gofmt to format your code (it's a good idea to get your IDE to format-on-save).

Also, this can be initialized with all the fields in one declaration:

```go
completeLaunchArgs := []string{
	"--idle-timeout-sec", string(int(sessionTimeout /time.Second)),
	"--auth-token-file", string(s.auth),
	"--android",
}
```

File: gapir/client/client.go (hunk 3 above)
Comment (purvisa-at-google-com) [reply gapir/client/client.go]: Android doesn't seem to like <> characters in intent names. I played around with escaping them with some success, but it seemed like an unnecessary risk supporting exotic characters for what seemed like a cosmetic choice.

File: gapir/client/session.go (hunk 4 above)
Comment (purvisa-at-google-com) [reply gapir/client/session.go]: Whoops. I do this in python all the time. It's a bad habit from being a c++ centric programmer.

Comment (purvisa-at-google-com) [reply gapir/client/session.go]: Good tip with the single initialiser. I only started go yesterday so while obvious, this didn't cross my mind.

File: cmd/gapir/cc/main.cpp (hunk 1 above)
Comment (purvisa-at-google-com) [reply cmd/gapir/cc/main.cpp]: --replay-archive does not do the same as --android.

I tried to find a way to express what the old android Parse() function was doing via arguments to the (now unified) PC Parse() function. Unfortunately I didn't see a way of achieving an exact behaviour match without adding a new flag.

File: core/cc/log.h (hunk 2 above)
Comment (purvisa-at-google-com) [reply core/cc/log.h]: Not unless we move them to (potentially) every c++ that uses log.h. This is quite a bit more than just Log.cpp

File: gapir/client/client.go (hunk 3 above)
Comment (ben-clayton) [reply gapir/client/client.go]: Sure - my question was more: why is this being passed down to adb at all? It's an internal key for `gapir/client`. It appears that you've had to change the value because you've started using this a string to pass to adb, which is now doing more than it was intended to.

Comment (purvisa-at-google-com) [reply gapir/client/client.go]: I can create a new key for the intent if you'd prefer to keep the <> characters. I just thought it was simpler to use fewer identifiers around the place.

File: cmd/gapir/cc/main.cpp (hunk 1 above)
Comment (hevrard) [reply cmd/gapir/cc/main.cpp]: What about passing a `struct Options *opts` to Parse(), such that you can pre-fill the options before parsing with different defaults depending on whether you're on host or device entry point?

File: gapir/client/client.go (hunk 3 above)
Comment (ben-clayton) [reply gapir/client/client.go]: There are other keys out there, none of them are reused for other string purposes, so I'd recommend just using the `"gapir-launch-args"` string literal at the site where you pass that at a parameter to the adb command, like all the other parameters.
My objection here is that there's nothing obvious or documented on this variable to say it is being used as a command line parameter string. Someone could decide they want to change the token value of `tyLaunchArgsKey` and unintentionally break the adb commands.

Comment (hevrard) [reply gapir/client/client.go]: This contant is mirrored in cmd/gapir/cc/main.cpp:596
```
const char* intent_flag = "gapir-launch-args";
```
This is fragile, as this string needs to exactly match the tyLaunchArgsKey.

I recommend to have a single declaration of the intent "keywork" as a different constant than tyLaunchArgsKey, and reuse that constant in both GAPIS and GAPIR, to make sure its actual value can be changed in a single place.

Comment (ben-clayton) [reply gapir/client/client.go]: > This is fragile, as this string needs to exactly match the tyLaunchArgsKey.

No. It should not have anything to do with `LaunchArgsKey`. :D

`gapir/client/session.go` has a whole bunch of command line parameter literal strings in it. Why is this one special?

If you create a whole bunch of named constants for command line parameters, okay - make them all constants, place them in `session.go`, but don't conflate it with device keys. They are completely unrelated concepts.

Comment (ben-clayton) [reply gapir/client/client.go]: To stress my point some more. `tyLaunchArgsKey` is a private type that is used purely as a unique map key with its single instance `LaunchArgsKey`. The only reason the underlying type is a string is so that it is easier to debug.
It could be an `int`, `struct`, `float64` - anything that produces a globally unique value-and-type key for the `bind.Registry.properties` map.
That's its soul purpose. Not to represent anything else.

File: cmd/gapir/cc/main.cpp (hunk 1 above)
Comment (purvisa-at-google-com) [reply cmd/gapir/cc/main.cpp]: Done!

File: gapir/client/client.go (hunk 3 above)
Comment (purvisa-at-google-com) [reply gapir/client/client.go]: Changed to a string literal as @hevrard suggested at my desk

File: cmd/gapir/cc/main.cpp [hunk 5]
```diff
@@ -76,6 +77,235 @@ std::vector<uint32_t> memorySizes {
...
+      } else if (strcmp(argv[i], "--auth-token-file") == 0) {
+        opts.SetMode(kReplayServer);
+        if (i + 1 >= argc) {
+          GAPID_FATAL("Usage: --auth-token-file <token-string>");
+        }
+        opts.authTokenFile = argv[++i];
+      } else if (strcmp(argv[i], "--enable-disk-cache") == 0) {
+        ensureNotAndroid(
+            "--enable-disk-cache");  // Fatal error if running on android. This
+                                     // flag is not supported.
```
[the remark this answers is not in the document]
Comment (purvisa-at-google-com) [reply cmd/gapir/cc/main.cpp]: Done

File: cmd/gapir/cc/main.cpp [hunk 6]
```diff
@@ -76,6 +77,235 @@ std::vector<uint32_t> memorySizes {
...
+        if (i + 1 >= argc) {
+          GAPID_FATAL("Usage: --disk-cache-path <cache-directory>");
+        }
+        opts.onDiskCacheOptions.path = argv[++i];
+      } else if (strcmp(argv[i], "--cleanup-on-disk-cache") == 0) {
+        ensureNotAndroid(
+            "--cleanup-on-disk-cache");  // Fatal error if running on android.
+                                         // This flag is not supported.
+        opts.onDiskCacheOptions.cleanUp = true;
+      } else if (strcmp(argv[i], "--port") == 0) {  // check it works
```
[the remark this answers is not in the document]
Comment (purvisa-at-google-com) [reply cmd/gapir/cc/main.cpp]: Done

File: cmd/gapir/cc/main.cpp [hunk 7]
```diff
@@ -395,8 +581,112 @@ void android_process(struct android_app* app, int32_t cmd) {
 
 }  // namespace
 
+// Extract command line arguments from the extra of Android intent:
+//   adb shell am start -n <...> -e gapir-launch-args "'list of arguments to be
```
[the remark this answers is not in the document]
Comment (purvisa-at-google-com) [reply cmd/gapir/cc/main.cpp]: @hevrard and I talked at my desk. "gapir" would get too many false positives in a grep of the source code; we settled on "gapir-intent-flag" in the end as this was believed to be unique enough to get only relevant hits.

File: gapir/client/client.go (hunk 3 above)
Comment (hevrard) [reply gapir/client/client.go]: Thanks Ben for the clarifications!

File: cmd/gapir/cc/main.cpp [hunk 8]
```diff
@@ -76,6 +77,190 @@ std::vector<uint32_t> memorySizes {
       128 * 1024 * 1024U,       // 128MB
 };
 
+struct Options {
```
Comment (AWoloszyn) [inline cmd/gapir/cc/main.cpp]: This Change is probably not the place to fix this, since I think it is a holdover from an earlier time. But none of these variables follow the Google C++ style guide (which we are at least TRYING to follow).

Comment (AWoloszyn) [reply cmd/gapir/cc/main.cpp]: https://google.github.io/styleguide/cppguide.html#General_Naming_Rules

File: cmd/gapir/cc/main.cpp [hunk 9]
```diff
@@ -395,8 +536,134 @@ void android_process(struct android_app* app, int32_t cmd) {
 
 }  // namespace
 
+// Extract command line arguments from the extra of Android intent:
+//   adb shell am start -n <...> -e gapir-launch-args "'list of arguments to be
+//   extracted'"
+// Note the quoting: from host terminal adb command, we need to double-escape
```
Comment (AWoloszyn) [inline cmd/gapir/cc/main.cpp]: We have run into some insane quoting things on windows so we should test there as well.

File: cmd/gapir/cc/main.cpp (hunk 8 above)
Comment (purvisa-at-google-com) [reply cmd/gapir/cc/main.cpp]: Yeah, I moved the code. The names are as they were when I encountered the source.

File: cmd/gapir/cc/main.cpp [hunk 10]
```diff
@@ -395,8 +536,134 @@ void android_process(struct android_app* app, int32_t cmd) {
...
+  std::vector<std::string> args;
+  args.push_back(
+      "gapir");  // POSIX says the first argv should be the program being run.
+                 // Lets inject a placeholder here for compatibility.
+
+  // Split extra_string
+  std::stringstream ss(extra_string);
+  std::string arg;
+
+  while (std::getline(ss, arg, ' ')) {
```
Comment (AWoloszyn) [inline cmd/gapir/cc/main.cpp]: This ignores quoted arguments. I am not sure if we want/need to support that, but things like spaces in filenames etc will break here.

File: cmd/gapir/cc/main.cpp (hunk 9 above)
Comment (purvisa-at-google-com) [reply cmd/gapir/cc/main.cpp]: This is copied from Graphics Fuzz.

File: cmd/gapir/cc/main.cpp (hunk 10 above)
Comment (purvisa-at-google-com) [reply cmd/gapir/cc/main.cpp]: I left this behaviour un-altered. The problem you highlight exists in the original code.

File: cmd/gapir/cc/main.cpp [hunk 11]
```diff
@@ -395,8 +536,134 @@ void android_process(struct android_app* app, int32_t cmd) {
 
 }  // namespace
 
+// Extract command line arguments from the extra of Android intent:
+//   adb shell am start -n <...> -e gapir-launch-args "'list of arguments to be
+//   extracted'"
+// Note the quoting: from host terminal adb command, we need to double-escape
+// the extra args string, as it is first quoted by host terminal emulator
+// (e.g. bash), then it must be quoted for the on-device shell.
+void GetArgvFromIntents(struct android_app* state, int* argc, char*** argv) {
```
Comment (AWoloszyn) [inline cmd/gapir/cc/main.cpp]: I won't hold up the CR for this, but I have a preference of making this slightly more C++-ey.
```
struct Args {
   int argc;
   char** argv;
   - Other stuff here?
   ~Args() { 
      - free argv here
   }
}
```

You can return by std::move(), so that we don't have to heap-alloc the struct itself.

Comment (AWoloszyn) [reply cmd/gapir/cc/main.cpp]: This way no-one has to forget to "Free"

File: cmd/gapir/cc/main.cpp [hunk 12]
```diff
@@ -821,7 +926,8 @@ static int startServer(core::CrashHandler* crashHandler, Options opts) {
 
 // Main function for PC
 int main(int argc, const char* argv[]) {
-  Options opts = Options::Parse(argc, argv);
+  Options opts;
```
Comment (AWoloszyn) [inline cmd/gapir/cc/main.cpp]: Any particular reason to switch this to passing by ref, and filling in?

Passing by ref violates the style guide as well. If you go down this route, you should pass by pointer. Style guide says you should only ever pass things by const-ref (which obviously won't work in this case)

Comment (AWoloszyn) [reply cmd/gapir/cc/main.cpp]:
```
Options opts = Options::Parse(argc, argv);
```
Should have RVO applied, so you shouldn't end up paying any cost for the "copy".

Comment (purvisa-at-google-com) [reply cmd/gapir/cc/main.cpp]: The pass by reference is not for "performance". It allows me to fill in some default values on android (that are gleaned from the apk) before later parsing the flags and applying the appropriate state changes on top.

I agree that copy elision should kick in here. And anyway the structure is so small and the operation once off, so optimizing it at the expense of clarity makes no sense.

Happily, I prefer passing things by pointer than non-cost reference. I used a reference at @hevrard  proposal. Would you prefer a pointer here?

Comment (AWoloszyn) [reply cmd/gapir/cc/main.cpp]: Understood!

We try our best to obey the style guide, so it would be better to make it a pointer.

File: cmd/gapir/cc/main.cpp (hunk 10 above)
Comment (AWoloszyn) [reply cmd/gapir/cc/main.cpp]: Sounds good. Feel free to leave this as-is.

File: cmd/gapir/cc/main.cpp (hunk 9 above)
Comment (AWoloszyn) [reply cmd/gapir/cc/main.cpp]: Alrighty. We should just make double-sure to test on windows before the next release.

File: core/cc/log.h (hunk 2 above)
Comment (purvisa-at-google-com) [reply core/cc/log.h]: Done!

File: cmd/gapir/cc/main.cpp (hunk 11 above)
Comment (purvisa-at-google-com) [reply cmd/gapir/cc/main.cpp]: Isn't this dangerous in its own right; that anyone returning Args who forgets to use std::move will return a pointer into freed memory?

Comment (purvisa-at-google-com) [reply cmd/gapir/cc/main.cpp]: OK, changed this code so it uses a std::vector<std::string> for data storage and string::c_str() / std::vector<unsigned  char*> to create a presentation of that data which looks like old fashioned c strings.

I think this is the most "c++ey" way of solving the problem.

God knows why anyone's still using c strings in c++.

Comment (AWoloszyn) [reply cmd/gapir/cc/main.cpp]: As for "anyone who forgets to use std::move" we can disable copy construction/copy assignment. Then they will just get a compile-error in this case.

Comment (purvisa-at-google-com) [reply cmd/gapir/cc/main.cpp]: What's your opinion of the std::string proposal I uploaded in the last revision?

Comment (AWoloszyn) [reply cmd/gapir/cc/main.cpp]: That being said how you restructured it is completely fine with me.

We typically use c strings in 2 cases.
1) When we don't really have a choice. (Interfacing with a C api. eg. `int main(....))`
2) When we want to avoid copies. std::string can be a pain for a number of reasons, and often we want to avoid copies. C++14 doesn't yet have string_view or equivalent, and we haven't pulled an implementation in yet.

But for most other cases raw cstrings should generally be avoided.

**Comment by AWoloszyn** [comment]
LGTM, you should just run clang-format before submitting to make the presubmit bots happy.
IIRC we use clang-format-6 on the bots. You can use the ./kokoro/presubmit/presubmit.sh to make sure everything is right. 

This only works if you have the same go/clang-format versions as our bots however, otherwise you can get differences.

File: cmd/gapir/cc/main.cpp (hunk 11 above)
Comment (AWoloszyn) [reply cmd/gapir/cc/main.cpp]: Yep the last version works for me. Thanks!
